Mathlib/Algebra/Category/Ring/Under.lean

```diff
@@ -0,0 +1,159 @@
+/-
+Copyright (c) 2024 Christian Merten. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Christian Merten
+-/
+import Mathlib.Algebra.Category.Ring.Colimits
+import Mathlib.Algebra.Category.Ring.Constructions
+import Mathlib.CategoryTheory.Adjunction.Over
+
+/-!
+# Under `CommRingCat`
+
+In this file we provide basic API for `Under R` when `R : CommRingCat`. `Under R` is
+(equivalent to) the category of commutative `R`-algebras. For not necessarily commutative
+algebras, use `AlgebraCat R` instead.
+-/
+
+noncomputable section
+
+universe u
+
+open TensorProduct CategoryTheory Limits
+
+namespace CommRingCat
+
+variable {R S : CommRingCat.{u}}
+
+instance : CoeSort (Under R) (Type u) where
+  coe A := A.right
+
+instance (A : Under R) : Algebra R A := RingHom.toAlgebra A.hom
+
+/-- Turn a morphism in `Under R` into an algebra homomorphism. -/
+def toAlgHom {A B : Under R} (f : A ⟶ B) : A →ₐ[R] B where
+  __ := f.right
+  commutes' a := by
+    have : (A.hom ≫ f.right) a = B.hom a := by simp
+    simpa only [Functor.const_obj_obj, Functor.id_obj, CommRingCat.comp_apply] using this
+
+@[simp]
+lemma toAlgHom_id (A : Under R) : toAlgHom (𝟙 A) = AlgHom.id R A := rfl
+
+@[simp]
+lemma toAlgHom_comp {A B C : Under R} (f : A ⟶ B) (g : B ⟶ C) :
+    toAlgHom (f ≫ g) = (toAlgHom g).comp (toAlgHom f) := rfl
+
+@[simp]
+lemma toAlgHom_apply {A B : Under R} (f : A ⟶ B) (a : A) :
+    toAlgHom f a = f.right a :=
+  rfl
+
+variable (R) in
+/-- -/
+@[simps! hom, simps! (config := .lemmasOnly) right]
+def mkUnder (A : Type u) [CommRing A] [Algebra R A] : Under R :=
+  Under.mk (CommRingCat.ofHom <| algebraMap R A)
+
+/-- Make a morphism in `Under R` from an algebra map. -/
+def fromAlgHom {A B : Type u} [CommRing A] [CommRing B] [Algebra R A] [Algebra R B]
+    (f : A →ₐ[R] B) : mkUnder R A ⟶ mkUnder R B :=
+  Under.homMk f.toRingHom <| by
+    ext a
+    exact f.commutes' a
+
+@[simp]
+lemma fromAlgHom_right_apply {A B : Type u} [CommRing A] [CommRing B] [Algebra R A] [Algebra R B]
+    (f : A →ₐ[R] B) (a : A) :
+    (fromAlgHom f).right a = f a :=
+  rfl
+
+@[simp]
+lemma fromAlgHom_comp {A B C : Type u} [CommRing A] [CommRing B] [CommRing C] [Algebra R A]
+    [Algebra R B] [Algebra R C] (f : A →ₐ[R] B) (g : B →ₐ[R] C) :
+    (fromAlgHom (g.comp f)) = fromAlgHom f ≫ fromAlgHom g :=
+  rfl
+
+lemma mkUnder_ext {A B : Type u} [CommRing A] [CommRing B] [Algebra R A] [Algebra R B]
```

Comment:
Should this be `@[ext]`? 
Is this not stated for general `A : Under R` so that the forall in `h` is over `A` and not `A.left` (or something else ugly)? If so, maybe `mkUnder R B` can still be some general `B : Under R`.